File: leetcode/20.有效的括号.ts
/*
 * @lc app=leetcode.cn id=20 lang=typescript
 *
 * [20] 有效的括号
 */

// @lc code=start
function isValid(s: string): boolean {
    const length = s.length;
    if (length % 2 !== 0) return false;

    const m = new Map([
        [')', '('],
        ['}', '{'],
        [']', '['],
    ])
    const tmp: string[] = [];
    for (let i = 0; i < length; i++) {
        if (tmp.length > 0 && tmp[tmp.length - 1] === m.get(s[i])) {
            tmp.pop();
        } else {
            tmp.push(s[i]);
        }
    }
    return tmp.length === 0;
};
// @lc code=end

File: leetcode/20.有效的括号.rs
/*
 * @lc app=leetcode.cn id=20 lang=rust
 *
 * [20] 有效的括号
 */

// @lc code=start
impl Solution {
    pub fn is_valid(s: String) -> bool {
        use std::collections::HashMap;
        let s = s.chars().collect::<Vec<char>>();
        let length = s.len();
        if length % 2 != 0 {
            return false;
        }
        let m: HashMap<_, _> = vec![(')', '('), ('}', '{'), (']', '[')]
            .into_iter()
            .collect();
        let mut tmp: Vec<char> = vec![];
        for c in s.iter() {
            match m.get(c) {
                Some(&val) if tmp.last() == Some(&val) => {
                    tmp.pop();
                }
                _ => {
                    tmp.push(*c);
                }
            }
        }
        tmp.is_empty()
    }
}
// @lc code=end
